onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -64,6 +68,210 @@ class CudnnReduceDescriptor final {
   cudnnReduceTensorDescriptor_t desc_;
 };
 
+// TODO ReduceKernel::ReduceKernelShared() is still used by some other training classes though it's not used here - this should be refactored.
+template <bool allow_multi_axes>
+template <typename T, typename OutT, cudnnReduceTensorIndices_t ReduceTensorIndices>
+Status ReduceKernel<allow_multi_axes>::ReduceKernelShared(
+    const T* X,
+    const TensorShape& input_shape,
+    OutT* Y,
+    const TensorShape& output_shape,
+    cudnnReduceTensorOp_t cudnnReduceOp,
+    std::vector<int64_t> output_dims) const {
+  typedef typename ToCudaType<T>::MappedType CudaT;
+  cudnnDataType_t cudnn_type_X = CudnnTensor::GetDataType<CudaT>();
+  const auto rank = input_shape.NumDimensions();
+
+// Block of fast matrix row reduction.
+// It relies on new atomicAdd for half type, so old CUDA can't use it.
+  const auto stride = input_shape[input_shape.NumDimensions() - 1];
```

Comment:
Should this use atomic_add from cuda/atomic/common.cuh?

If not, what happens if it's an old CUDA version? I didn't notice any check on __CUDA_ARCH__ that would provide a compile-time or runtime error message saying why it's broken.

Reply:
Yes, it should use our own `atomic_add`.

Reply:
So this comment no longer applies and should be removed? 
// It relies on new atomicAdd for half type, so old CUDA can't use it.

Reply:
Right. After switching to `atomic_add`, this comment is no longer valid.

Reply:
will remove